ToolsPanel: Use Grid component to handle layout

aaronrobertshaw opened the pull request:
Related:
- https://github.com/WordPress/gutenberg/pull/35415
- https://github.com/WordPress/gutenberg/issues/35058

Follow-up:
- https://github.com/WordPress/gutenberg/pull/35911

## Description

This PR refactors the `ToolsPanel` to leverage the `Grid` component to handle its layout internally.

By adopting the `Grid` component to manage the `ToolsPanel` layout, the panel will be better positioned to expose any custom grid props we desire in future. It will also remove the reliance on the currently hard-coded `single-column` class name.

## How has this been tested?

Manually. 

1. Ensured the layout and behaviour of the `ToolsPanel` in the storybook examples
2. Ensured the `ToolsPanel` displays block support panels and controls correctly.
    - Typography panel will not apply single column span until styles are tweaked in a follow-up
3. Adjusted various panels' and controls' props to play with custom column layouts.

#### Testing instructions

1. Checkout this branch and fire up the storybook: `npm run storybook:dev`
2. In the page that opens, select the [ToolsPanel > Default](http://localhost:50240/?path=/story/components-experimental-toolspanel--default) example from the left sidebar
3. Ensure the panel is laid out correctly and behaves
4. Switch to the block editor, create a post, add a cover block and select it
5. Ensure that the "Dimensions" panel in the inspector controls displays appropriately

## Screenshots <!-- if applicable -->

<img width="294" alt="Screen Shot 2021-10-25 at 3 17 27 pm" src="https://user-images.githubusercontent.com/60436221/138638454-c937ca02-656e-4f90-9063-1fdb46bdbac4.png">

## Types of changes
New feature.

The removal of the reliance on the `single-column` class is a breaking change although the component is still experimental and the only adoption of the `ToolsPanel` that has merged is the "Dimensions" panel that does not use `single-column`. 

**_The Typography panel has now been switched to the ToolsPanel as well however that change hasn't been included in a release yet_**

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

aaronrobertshaw committed:
Switch ToolsPanel to use grid component

aaronrobertshaw committed:
Use new gridColumn prop for single column controls

aaronrobertshaw committed:
Add story illustrating custom panel columns

aaronrobertshaw commented on packages/components/src/tools-panel/styles.ts:
```diff
@@ -14,24 +14,19 @@ import { COLORS, CONFIG } from '../utils';
-		`,
 		fullWidth: css`
-			grid-column: span 2;
+			grid-column: 1 / -1;
```
The prevailing use-case for controls within a `ToolsPanel` will be to span the full width. This is aimed at making panel items span all columns by default.

aaronrobertshaw commented on packages/components/src/tools-panel/tools-panel-item/README.md:
```diff
@@ -18,6 +18,13 @@ for how to use `ToolsPanelItem`.
+how many columns it should span. This maps to the CSS `grid-column` property.
+
+- Required: No
+
```
I opted to mirror the CSS `grid-column` property here such that we might get the option to specify a specific column or span for free.

ciampo replied:
Makes sense. I also think that we should potentially look into creating a `GridItem` component, mirroring how the `Flex` component has a `FlexItem` component (to be done in a separate PR)

aaronrobertshaw commented on packages/components/src/tools-panel/tools-panel-item/hook.ts:
```diff
@@ -117,8 +123,15 @@ export function useToolsPanelItem(
 			styles.ToolsPanelItemPlaceholder;
-		return cx( styles.ToolsPanelItem, placeholderStyle, className );
-	}, [ isShown, shouldRenderPlaceholder, className ] );
+		const columnClass = css( { gridColumn } );
```
I'm not sure if this is the best approach to allow individual panel items to override their `grid-column` CSS. Let me know if there is a cleaner option.

andrewserong reviewed (commented):
Nice work, @aaronrobertshaw, this looks great in the Storybook demo, I think this is going to allow for some really nice layouts. It's quite fun to interact with:

![Kapture 2021-10-15 at 12 07 08](https://user-images.githubusercontent.com/14988353/137416389-9d3be0fe-28b8-40c8-a0a6-797603e4c313.gif)

I like the addition of being able to set `columns` at the ToolsPanel level. While I imagine we won't be refactoring away from Grid, I was wondering if using `gridColumn` at the ToolsPanelItem level ties the prop closely to the specific implementation? I could imagine a plugin author going to add a ToolsPanelItem to an existing panel and not being sure what value to put into the `gridColumn` string. Would an integer based `columnSpan` prop be clearer as an API? Also, if we ever needed to refactor to `flex`, the integer value could be used to calculate flex-basis without the `ToolsPanelItem` needing to change its value.

Of course, on the other hand, the `gridColumn` prop passing the value directly to the CSS does give the ToolsPanelItem more control if someone wanted to set a different `grid-column-start` / `grid-column-end` value than just spanning, so keeping it as is could offer more flexibility.

Curious to hear what you and other folks think, but this is testing nicely for me in Storybook and in existing ToolsPanels in the editor 👍

aaronrobertshaw commented:
Thanks for taking this for a test drive 👍 

> Of course, on the other hand, the gridColumn prop passing the value directly to the CSS does give the ToolsPanelItem more control if someone wanted to set a different grid-column-start / grid-column-end value than just spanning, so keeping it as is could offer more flexibility.

This is why I went with the current approach. More power and flexibility for free and it aligns more directly with an external spec i.e. CSS.

> and not being sure what value to put into the gridColumn string

I mentioned the direct mapping to the CSS property in the README so it should be clear. I originally had a link to external CSS specs but couldn't find many other cases of doing that so removed it.

> Would an integer based columnSpan prop be clearer as an API? Also, if we ever needed to refactor to flex, the integer value could be used to calculate flex-basis without the ToolsPanelItem needing to change its value.

I did consider it however it seemed to be unnecessarily restrictive when we could get the positioning functionality for free otherwise. If we needed to refactor to flex `columnSpan` would no longer make sense either as a name.

andrewserong commented:
> I did consider it however it seemed to be unnecessarily restrictive when we could get the positioning functionality for free otherwise

Thanks for explaining the rationale! Agreed, since it offers better flexibility and matches the CSS spec, that sounds like a good way to go. And it avoids adding an extra abstraction with the columnSpan idea that didn’t really map to the right CSS property anyway 👍

ciampo commented on packages/components/src/tools-panel/tools-panel/README.md:
```diff
@@ -67,6 +67,14 @@ export function DimensionPanel( props ) {
 
 ## Props
 
+### `columns`: `number`
```
We should probably remove this prop from here and instead add a `Inherited` props section at the end

```markdown
### Inherited props

`ToolsPanel` also inherits all of the [`Grid` props](/packages/components/src/grid/README.md#props).
```

ciampo commented on packages/components/src/tools-panel/styles.ts:
```diff
@@ -14,24 +14,19 @@ import { COLORS, CONFIG } from '../utils';
 
 export const ToolsPanel = css`
-	${ toolsPanelGrid.container };
+	${ toolsPanelGrid.spacing };
```
I think that a better solution here would be to remove the `spacing` styles altogether.

We should use the `columnGap` and `rowGap` props for the `Grid` item, and pass down those same props to `ToolsPanelWithInnerWrapper` and used them for CSS styles there.

aaronrobertshaw replied:
This will be addressed in a follow-up where we will decide just what to expose or add as new props to the existing components. For now, the [idea](https://github.com/WordPress/gutenberg/pull/35621#issuecomment-946106588) is to [only use the Grid internally](https://github.com/WordPress/gutenberg/pull/35621#issuecomment-946071523) to the `ToolsPanel`.

ciampo commented on packages/components/src/tools-panel/types.ts:
```diff
@@ -100,6 +105,12 @@ export type ToolsPanelItemProps = ToolsPanelItem & {
+	 * grid or how many columns it should span. This maps to the CSS grid-column
+	 * property.
+	 */
+	gridColumn?: string;
```
A better way to type CSS-related properties is to use the `CSSProperties` type:

```diff
diff --git a/packages/components/src/tools-panel/types.ts b/packages/components/src/tools-panel/types.ts
index 0d48bfb36a..ed3a62af12 100644
--- a/packages/components/src/tools-panel/types.ts
+++ b/packages/components/src/tools-panel/types.ts
@@ -2,7 +2,7 @@
  * External dependencies
  */
 // eslint-disable-next-line no-restricted-imports
-import type { ReactNode } from 'react';
+import type { ReactNode, CSSProperties } from 'react';
 
 type ResetAllFilter = () => void;
 type ResetAll = ( filters?: ResetAllFilter[] ) => void;
@@ -110,7 +110,7 @@ export type ToolsPanelItemProps = ToolsPanelItem & {
 	 * grid or how many columns it should span. This maps to the CSS grid-column
 	 * property.
 	 */
-	gridColumn?: string;
+	gridColumn?: CSSProperties[ 'gridColumn' ];
 	/**
 	 * Called when this item is deselected in the `ToolsPanel` menu. This is
 	 * normally used to reset the panel item control's value.
```

We should also reflect those changes in the README

aaronrobertshaw replied:
✅ Done. Thanks.

ciampo commented on packages/components/src/tools-panel/types.ts:
```diff
@@ -12,6 +12,11 @@ export type ToolsPanelProps = {
+	 * Definition of columns for the ToolsPanel's grid layout. Also used to
+	 * default panel items to span the full panel's width.
+	 */
+	columns: number;
```
Since we're using `Grid` internally, we could extend `Grid`'s types:

```diff
diff --git a/packages/components/src/tools-panel/types.ts b/packages/components/src/tools-panel/types.ts
index 0d48bfb36a..04be622acd 100644
--- a/packages/components/src/tools-panel/types.ts
+++ b/packages/components/src/tools-panel/types.ts
@@ -4,19 +4,19 @@
 // eslint-disable-next-line no-restricted-imports
 import type { ReactNode } from 'react';
 
+/**
+ * Internal dependencies
+ */
+import type { Props as GridProps } from '../grid/types';
+
 type ResetAllFilter = () => void;
 type ResetAll = ( filters?: ResetAllFilter[] ) => void;
 
-export type ToolsPanelProps = {
+export type ToolsPanelProps = GridProps & {
 	/**
 	 * The child elements.
 	 */
 	children: ReactNode;
-	/**
-	 * Definition of columns for the ToolsPanel's grid layout. Also used to
-	 * default panel items to span the full panel's width.
-	 */
-	columns: number;
 	/**
 	 * Flags that the items in this ToolsPanel will be contained within an inner
 	 * wrapper element allowing the panel to lay them out accordingly.
```

We should just check all the props that come with [`GridProps`](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/grid/types.ts#L26) and see if there's anything that we don't want to accept in `ToolsPanel`

aaronrobertshaw replied:
I've tried applying this patch as well as making required updates to the typing of `ToolsPanelWithInnerWrapper` in `styles.ts`. I still end up with a type error for the `useToolsPanel` hook. It appears that it now fails to infer the type, possibly due to the now larger number of possibilities for `otherProps`?

My lack of TypeScript knowledge is a bit of a problem here. I'm hoping you can point me in the right direction to address this one.

<details>
  <summary>Here are the changes I have before the point I tried typing the `useToolsPanel` return value</summary>
  
  ```diff
diff --git a/packages/components/src/tools-panel/styles.ts b/packages/components/src/tools-panel/styles.ts
index d9aa409abd..f986291c4d 100644
--- a/packages/components/src/tools-panel/styles.ts
+++ b/packages/components/src/tools-panel/styles.ts
@@ -12,6 +12,7 @@ import {
 } from '../base-control/styles/base-control-styles';
 import { COLORS, CONFIG } from '../utils';
 import { space } from '../ui/utils/space';
+import { useResponsiveValue } from '../ui/utils/use-responsive-value';
 
 const toolsPanelGrid = {
 	spacing: css`
@@ -39,11 +40,16 @@ export const ToolsPanel = css`
  * CSS grid display to be re-established.
  */
 
-export const ToolsPanelWithInnerWrapper = ( columns: number ) => {
+export const ToolsPanelWithInnerWrapper = (
+	columns: Array< number | undefined > | number
+) => {
+	const columnsAsArray = Array.isArray( columns ) ? columns : [ columns ];
+	const column = useResponsiveValue( columnsAsArray );
+
 	return css`
 		> div:not( :first-of-type ) {
 			display: grid;
-			grid-template-columns: ${ `repeat( ${ columns }, 1fr )` };
+			grid-template-columns: ${ `repeat( ${ column }, 1fr )` };
 			${ toolsPanelGrid.spacing }
 			${ toolsPanelGrid.item.fullWidth }
 		}
diff --git a/packages/components/src/tools-panel/types.ts b/packages/components/src/tools-panel/types.ts
index ed3a62af12..c91634eab5 100644
--- a/packages/components/src/tools-panel/types.ts
+++ b/packages/components/src/tools-panel/types.ts
@@ -4,19 +4,15 @@
 // eslint-disable-next-line no-restricted-imports
 import type { ReactNode, CSSProperties } from 'react';
 
+/**
+ * Internal dependencies
+ */
+import type { Props as GridProps } from '../grid/types';
+
 type ResetAllFilter = () => void;
 type ResetAll = ( filters?: ResetAllFilter[] ) => void;
 
-export type ToolsPanelProps = {
-	/**
-	 * The child elements.
-	 */
-	children: ReactNode;
-	/**
-	 * Definition of columns for the ToolsPanel's grid layout. Also used to
-	 * default panel items to span the full panel's width.
-	 */
-	columns: number;
+export type ToolsPanelProps = GridProps & {
 	/**
 	 * Flags that the items in this ToolsPanel will be contained within an inner
 	 * wrapper element allowing the panel to lay them out accordingly.
```

</details>

aaronrobertshaw replied:
We'll hold off on any changes along these lines until we arrive at a decision regarding what we wish to expose in terms of props later on for the `ToolsPanel`.

ciampo reviewed (commented):
Thank you for continuing to improve the `ToolsPanel` component — the new grid layout looks great!

mirka reviewed (commented):
I'm wondering what happens in cases like this, where you have an implicit group of optional panel items that belong in their own row. For example width/height/depth and padding/margin. The "wrap" behavior ("auto-placement"? unsure what the correct term is!) is not what you want:

https://user-images.githubusercontent.com/555336/137523775-d8ca1cc9-51b7-48f8-b33c-359bdca38d12.mp4

Setting explicit `grid-column-start` values will make it better, but then it doesn't auto-place within the row:

https://user-images.githubusercontent.com/555336/137528883-1535be25-fd2f-4be3-93de-28ddfcae7a34.mp4

I'm probably missing context on how ToolsPanel is intended to be used, but I was actually imagining the consumer being responsible for doing the layout (`Grid` or otherwise) inside ToolsPanel.

```jsx
<ToolsPanel>
  <Grid columns={2} gap={2}>
    <ToolsPanelItem />
    <ToolsPanelItem />
  </Grid>
  <Grid columns={3} gap={2}>
    <ToolsPanelItem />
    <ToolsPanelItem />
    <ToolsPanelItem />
  </Grid>
</ToolsPanel>
```

One could also say that an implicit group (like width/height/depth) should be encapsulated inside a single ToolsPanelItem. In that case, it might be that an elaborate `gridColumn` system built into ToolsPanel is a bit overkill, because there is a limit to the amount of random auto-placement you want to occur on unrelated optional items that are not full-width.

Just a non-blocking observation, please disregard if it doesn't matter 🙂

aaronrobertshaw committed:
Improve ToolsPanel README around inherited Grid props

aaronrobertshaw committed:
Update typing of gridColumn to use CSSProperties

aaronrobertshaw commented:
Thanks for the feedback @mirka 👍 

I'll explore the suggestion further and see what I can come up with.

> I'm wondering what happens in cases like this, where you have an implicit group of optional panel items that belong in their own row. For example width/height/depth and padding/margin. The "wrap" behavior ("auto-placement"? unsure what the correct term is!) is not what you want:

I believe the idea was for all the controls within a `ToolsPanel` to be related in a way that such implicit groups wouldn't really be needed. In your example, this would be the separation of the dimensions vs spacing controls into separate panels. This has been a [topic of discussion](https://github.com/WordPress/gutenberg/issues/28356#issuecomment-899587151) already although I'm not sure if we've settled on anything concrete yet.

Perhaps @jasmussen might have some insight as to what the `ToolsPanel` really needs to achieve in terms of layout in the bigger picture?

> I'm probably missing context on how ToolsPanel is intended to be used, but I was actually imagining the consumer being responsible for doing the layout (Grid or otherwise) inside ToolsPanel.

At a glance, this would provide more flexibility though I'm not sure how it would play when SlotFills are used to allow injection of controls into the panel. We'd likely be back to adding a generic grid wrapping all those fills and this same limitation/issue. This is something I'll need to dig into a little deeper.

> One could also say that an implicit group (like width/height/depth) should be encapsulated inside a single ToolsPanelItem. 

I'm not sure I agree with this. These individual items need to be represented within, display controlled, or value reset, by the tools panel menu.

The current approach in the `ToolsPanel` allows an individual `ToolsPanelItem` to register itself with the panel via the `ToolsPanelContext`. This does allow items to be wrapped in other components etc. As a result, the current functionality might already be close to supporting the approach within your snippet.

If it turns out to be more involved, given this was non-blocking as you mentioned, a follow-up PR refining this layout could be an option.

jasmussen commented:
The neat arrangement of controls in the inspector, as more are added or removed, is definitely a tricky one, and important one to get right. In many ways, that's a key part of addressing https://github.com/WordPress/gutenberg/issues/27331: ensuring a system where controls that get shown or hidden as part of theme.json or global styles or users, stil arranges itself in a neat way, and can appear consistent across inspectors.

Ultimately I'm not personally sure how far we can go with regards to controls resizing and adjusting themselves depending on context, I'll need to defer to you all (and CC: @ciampo as well). 

But just as a baseline, I imagine controls registering themselves as half or full-size controls, possibly with an option for 3rds as well, and then as one half-size item gets hidden, it will make room for a subsequent half-size item to take its place:
<img width="688" alt="Frame 862" src="https://user-images.githubusercontent.com/1204802/137683874-70116ef2-cfb5-4db6-bdb0-53d3cdfad50f.png">

It doesn't seem like it will be necessary for a full size control to be able to shift between full-size and half-size _automatically_, but it's an optimization that could be looked at if/when it becomes necessary.

aaronrobertshaw commented:
Thanks for sharing that @jasmussen 👍 

> But just as a baseline, I imagine controls registering themselves as half or full-size controls, possibly with an option for 3rds as well, and then as one half-size item gets hidden, it will make room for a subsequent half-size item to take its place:

This sounds to me like the current behaviour. The `ToolsPanel` was built to satisfy the mockups in https://github.com/WordPress/gutenberg/issues/27331.

I think the question Lena is asking is whether that next subsequent half-size item should in fact take its place in some situations. 

---

> This does allow items to be wrapped in other components etc. As a result, the current functionality might already be close to supporting the approach within your snippet.

After a little testing, I believe we can achieve the ["implicit group layout" raised earlier](https://github.com/WordPress/gutenberg/pull/35621#pullrequestreview-781088684) with only a small tweak to the styles to hide an empty grid. The items in the group would be wrapped in a `Grid` similar to the suggested markup.


https://user-images.githubusercontent.com/60436221/137690582-b0d15174-9522-4b48-b0e1-c806a8337b3d.mp4

<details>
  <summary>Here's a diff for the style tweak and storybook example changes demo'd in the above video.</summary>
  
```diff
diff --git a/packages/components/src/tools-panel/stories/index.js b/packages/components/src/tools-panel/stories/index.js
index e9a39b2932..e5b7d9b73e 100644
--- a/packages/components/src/tools-panel/stories/index.js
+++ b/packages/components/src/tools-panel/stories/index.js
@@ -15,6 +15,7 @@ import { ToolsPanel, ToolsPanelItem } from '../';
 import Panel from '../../panel';
 import UnitControl from '../../unit-control';
 import { createSlotFill, Provider as SlotFillProvider } from '../../slot-fill';
+import { Grid } from '../../grid';
 
 export default {
 	title: 'Components (Experimental)/ToolsPanel',
@@ -258,45 +259,49 @@ export const WithCustomColumns = () => {
 							onChange={ ( next ) => setSize( next ) }
 						/>
 					</ToolsPanelItem>
-					<ToolsPanelItem
-						gridColumn="span 1"
-						hasValue={ () => !! width }
-						label="Width"
-						onDeselect={ () => setWidth( undefined ) }
-						isShownByDefault={ true }
+					<Grid
+						style={ { gridColumn: '1/-1' } }
+						columns={ 3 }
+						columnGap={ 16 }
+						rowGap={ 24 }
 					>
-						<UnitControl
+						<ToolsPanelItem
+							gridColumn="span 1"
+							hasValue={ () => !! width }
 							label="Width"
-							value={ width }
-							onChange={ ( next ) => setWidth( next ) }
-						/>
-					</ToolsPanelItem>
-					<ToolsPanelItem
-						gridColumn="span 1"
-						hasValue={ () => !! height }
-						label="Height"
-						onDeselect={ () => setHeight( undefined ) }
-						isShownByDefault={ true }
-					>
-						<UnitControl
+							onDeselect={ () => setWidth( undefined ) }
+						>
+							<UnitControl
+								label="Width"
+								value={ width }
+								onChange={ ( next ) => setWidth( next ) }
+							/>
+						</ToolsPanelItem>
+						<ToolsPanelItem
+							gridColumn="span 1"
+							hasValue={ () => !! height }
 							label="Height"
-							value={ height }
-							onChange={ ( next ) => setHeight( next ) }
-						/>
-					</ToolsPanelItem>
-					<ToolsPanelItem
-						gridColumn="span 1"
-						hasValue={ () => !! depth }
-						label="Depth"
-						onDeselect={ () => setDepth( undefined ) }
-						isShownByDefault={ true }
-					>
-						<UnitControl
+							onDeselect={ () => setHeight( undefined ) }
+						>
+							<UnitControl
+								label="Height"
+								value={ height }
+								onChange={ ( next ) => setHeight( next ) }
+							/>
+						</ToolsPanelItem>
+						<ToolsPanelItem
+							gridColumn="span 1"
+							hasValue={ () => !! depth }
 							label="Depth"
-							value={ depth }
-							onChange={ ( next ) => setDepth( next ) }
-						/>
-					</ToolsPanelItem>
+							onDeselect={ () => setDepth( undefined ) }
+						>
+							<UnitControl
+								label="Depth"
+								value={ depth }
+								onChange={ ( next ) => setDepth( next ) }
+							/>
+						</ToolsPanelItem>
+					</Grid>
 					<ToolsPanelItem
 						gridColumn="span 2"
 						hasValue={ () => !! padding }
diff --git a/packages/components/src/tools-panel/styles.ts b/packages/components/src/tools-panel/styles.ts
index d9aa409abd..405a621510 100644
--- a/packages/components/src/tools-panel/styles.ts
+++ b/packages/components/src/tools-panel/styles.ts
@@ -31,6 +31,10 @@ export const ToolsPanel = css`
 	border-top: ${ CONFIG.borderWidth } solid ${ COLORS.gray[ 200 ] };
 	margin-top: -1px;
 	padding: ${ space( 4 ) };
+
+	> div:empty {
+		display: none;
+	}
 `;
 
 /**
```
</details>


If this satisfies both giving consumers the ability to group controls while still allowing a default layout to be applied by the `ToolsPanel` thereby reducing friction from a devx perspective I think we could move forward with this approach without any drastic changes.

I'm open to other ideas or thoughts in case I've missed something.

jasmussen commented:
> I think the question Lena is asking is whether that next subsequent half-size item should in fact take its place in some situations.

Yes, I'm missing some nuance. For example, I'm not sure we should introduce thirds yet — it's possible it might be necessary in the future, but it doesn't seem like something to optimize for, especially yet. If we can accomplish the designs in #34574, that feels like it will put us in a good place.

But more specifically, I don't know that we'll need to optimize for a design like this:

<img src="https://user-images.githubusercontent.com/60436221/137276744-ee707c85-4aa0-48e8-9ec0-c6c35ed279e5.png">

Especially for the tools panel, we'll curate the sequence of panels to always be in the same sequence. For example, full-width controls will almost always be at the top, with half size controls below that. So you shouldn't encounter a situation where you remove a half-size item, and a full size item can't take it's place.

So to rephrase it a bit: the panels need smarts so that when items are hidden or added, controls reorient themselves. However we'll nearly always want to curate the _sequence_ of items.

ciampo commented:
Sorry for being a bit late to the conversation! I haven't really got the the time to review the code patches in depth, and from tomorrow I won't be able to provide feedback on GitHub reliably until November 8.

Having said that, I believe that a good compromise and a _safer_ approach would be to keep things simple for now — we can always add more complex layout behaviours later. In contrast, adding stuff now that we may need to remove later on will be more disruptive. The approach that I propose for the changes _in this PR_ is:

- refactor the component to use `Grid` internally, and keep the current 2 columns layout implementation
- remove the `single-column` hardcoded className
- Do not add any new prop on `ItemGroup` — for now, we may agree that, by default, each `ToolsPanelItem` occupies 1 columns in the layout. If that needs to be changed, the user of `ToolsPanelItem` may want to apply CSS grid styles externally via `className` (e.g. by adding `grid-column: 1 / -1` for full-width)

This solution buys us some time to discuss around a few questions that come to mind, which may need further time and exploration before reaching an agreement:

- How many columns should the `ToolsPanel` layout have? (e.g 2, 3...) Should the number of columns be exposed as a prop? Should it changed based on the width of the component (e.g. using container queries)?
- Should we add additional props to `ToolsPanelItem` (e.g. `gridColumn`)? Should these props be a 1:1 with the CSS properties?
- How do we handle the potential auto-placing algorithm of CSS grid when hiding / showing elements?

What do folks think? (Also note, as said earlier, that I may not be able to follow up to this comment in the next days, until November 8)

mirka commented:
> If this satisfies both giving consumers the ability to group controls while still allowing a default layout to be applied by the `ToolsPanel` thereby reducing friction from a devx perspective I think we could move forward with this approach without any drastic changes.

@aaronrobertshaw Nice, this addresses my concern perfectly, and it looks hopeful that we can have the flexibility while providing a default layout!

That said, Marco's plan of not exposing new props in this PR is probably wise. It warrants some more exploration/discussion, and I'm still doubtful that we need to expose unrealistic (>2 columns) or "implementation detail" (`gridColumn`) options as part of ToolsPanel.

aaronrobertshaw committed:
Only use Grid internally and remove recent prop additions

aaronrobertshaw committed:
Remove obsolete inherited props section from readme

aaronrobertshaw commented:
> - refactor the component to use Grid internally, and keep the current 2 columns layout implementation
> - remove the single-column hardcoded className
> - Do not add any new prop on ItemGroup — for now, we may agree that, by default, each ToolsPanelItem occupies 1 columns in the layout. If that needs to be changed, the user of ToolsPanelItem may want to apply CSS grid styles externally via className (e.g. by adding grid-column: 1 / -1 for full-width)

This PR has been updated to address the majority of @ciampo's proposal. The main difference is that items still default to spanning the full width. This is by far the most common use case for the `ToolsPanel` to date and I believe requires consumers to make the least amount of custom styling changes.

### Action Taken

- Removed the columns prop from `ToolsPanel`
- Prevented blindly passing props through to the `Grid` component used as the root element for the panel
    - This was to avoid introducing extra markup and styling changes if moving the grid to an inner component
- Removed `gridColumn` prop from the `ToolsPanelItem`. Handling grid column styling is now up to consumers
- Kept the `ToolsPanel` defaulting to 2 columns
- Kept `ToolsPanelItem`s defaulting to spanning all columns
- Removed the custom columns example from the storybook
- Updated other storybook examples to handle single column span styles
- Updated types according to the above changes
- Updated docs

### Summary

After making the above changes:

- `ToolsPanel` is behaving as expected for me in the storybook
- Unit tests are all green
- Type checking passes
- `ToolsPanel` works for block supports in the inspector controls
- SlotFill provided items are still rendering correctly within their inner wrapper

I do think all this begs the question though as to whether this PR should be put on ice until the required discussions are had and a decision on behaviour made. As it is now, we're switching to the grid component to handle layout but not really allowing it to do any of that. 

Given this makes very little in the way of user-facing or consumer-orientated changes, would it be better to leave the status quo so this is iterated on once properly?

aaronrobertshaw commented:
Having had some time to reflect, I think it would be best to land this so that the reliance on the `ToolsPanel` providing the `single-column` class is removed. 

The recent switch of the [Typography block supports panel to use the `ToolsPanel`](https://github.com/WordPress/gutenberg/pull/33744) relies on some controls only spanning a single column. I'll create a follow up to this PR that will re-establish the desired layout in the Typography panel.

aaronrobertshaw committed:
Undo whitespace change

aaronrobertshaw committed:
Tweak styling in storybook

mirka commented on packages/components/src/tools-panel/stories/index.js:
```diff
@@ -224,6 +224,14 @@ export const WithSlotFillItems = () => {
+
+	.single-column {
+		grid-column: span 1;
+	}
```
Not super important since our columns API is still in flux, but since story code serve as examples it might be nicer to have a `styled( ToolsPanelItem )` rather than use an explicit class name here. (Non-blocking nit)

aaronrobertshaw replied:
Thanks for the suggestion @mirka and your time reviewing this PR. 

I've just pushed a [commit](https://github.com/WordPress/gutenberg/pull/35621/commits/b145d5b9e47e722b60385be8ada64304fd427364) switching to a styled component for the stories as well as a [changelog](https://github.com/WordPress/gutenberg/pull/35621/commits/102ac1315e195f8518f85050a315fb221ebef589).

If you're still happy with the changes, I can merge this after the e2es pass.

mirka replied:
Let's do it! 🚢

mirka reviewed (approved):
> Having had some time to reflect, I think it would be best to land this so that the reliance on the `ToolsPanel` providing the `single-column` class is removed.

Sounds good, and thanks for the solid improvements and exploration here 👍 

Let's also add a brief entry under the "Experimental" section in the changelog.

aaronrobertshaw commented:
A follow-up to fix the Typography panel layout and letter-spacing control after this PR lands can be found in https://github.com/WordPress/gutenberg/pull/35911. It would be best to land these as close together as possible.

aaronrobertshaw committed:
Update single column examples to used styled component

aaronrobertshaw committed:
Add changelog

ciampo commented on packages/components/src/tools-panel/tools-panel/component.tsx:
```diff
@@ -24,11 +24,14 @@ const ToolsPanel = (
 		panelContext,
 		resetAllItems,
 		toggleItem,
-		...toolsPanelProps
```
I believe that removing this line can potentially introduce breaking changes, since we are not passing extra props to the wrapper component (e.g. any standard HTML props, like `data-*` attributes, `role`, `aria-*` props...)

We should probably reintroduce the spreading of those extra props to the `Grid` wrapper.

aaronrobertshaw replied:
Agreed, the removal of that was probably an overreach in response to the request to not expose any props on the `ToolsPanel` until they had been considered more.

To reinstate the collection, and subsequent spreading, of additional props for the `ToolsPanel` I ran into problems being able to appease TypeScript while attempting to exclude `Grid` props. If you have any ideas on the best way to achieve that, I'd appreciate your input.

In the meantime, I could easily clear any `Grid` props before they are passed through to the `ToolsPanel` component. I've included a diff below to illustrate what I'm meaning. If that happens to be acceptable, I'll create a PR for it.

<details>
  <summary>Patch to clear Grid props</summary>
  
```diff
diff --git a/packages/components/src/tools-panel/tools-panel/component.tsx b/packages/components/src/tools-panel/tools-panel/component.tsx
index 5492e19341..6525a62d3c 100644
--- a/packages/components/src/tools-panel/tools-panel/component.tsx
+++ b/packages/components/src/tools-panel/tools-panel/component.tsx
@@ -24,14 +24,14 @@ const ToolsPanel = (
 		panelContext,
 		resetAllItems,
 		toggleItem,
-		className,
+		...toolsPanelProps
 	} = useToolsPanel( props );
 
 	// Props are not directly passed through to avoid exposing Grid props
 	// until agreement has been reached on how ToolsPanel layout should be
 	// handled.
 	return (
-		<Grid columns={ 2 } className={ className } ref={ forwardedRef }>
+		<Grid { ...toolsPanelProps } columns={ 2 } ref={ forwardedRef }>
 			<ToolsPanelContext.Provider value={ panelContext }>
 				<ToolsPanelHeader
 					label={ label }
diff --git a/packages/components/src/tools-panel/tools-panel/hook.ts b/packages/components/src/tools-panel/tools-panel/hook.ts
index 2cedad4e73..2efd3b3121 100644
--- a/packages/components/src/tools-panel/tools-panel/hook.ts
+++ b/packages/components/src/tools-panel/tools-panel/hook.ts
@@ -216,5 +216,19 @@ export function useToolsPanel(
 		resetAllItems,
 		toggleItem,
 		className: classes,
+		// The following properties are cleared so as to prevent `Grid` props
+		// being passed through. This is intended to be removed once it is
+		// decided on exactly which `Grid` props the `ToolsPanel` should expose.
+		align: undefined,
+		alignment: undefined,
+		columns: undefined,
+		columnGap: undefined,
+		isInline: undefined,
+		gap: undefined,
+		justify: undefined,
+		rowGap: undefined,
+		rows: undefined,
+		templateColumns: undefined,
+		templateRows: undefined,
 	};
 }

```
</details>

ciampo replied:
At the moment, the fact that `ToolsPanel` uses `Grid` internally is an implementation detail. We're not exposing that in any APIs, docs or types. Therefore, I don't think that we need to explicitly exclude `Grid`-related props, as we never included them in the first place.

Sure, that won't stop a developer from passing those props to `ToolsPanel` at runtime — but that's not the correct, documented usage anyway.

<details>

<summary>I think we should just bring back the spreading of `toolsPanelProps` without worrying too much about the rest:</summary>

```diff
diff --git a/packages/components/src/tools-panel/tools-panel/component.tsx b/packages/components/src/tools-panel/tools-panel/component.tsx
index 5492e19341..f7f16240ce 100644
--- a/packages/components/src/tools-panel/tools-panel/component.tsx
+++ b/packages/components/src/tools-panel/tools-panel/component.tsx
@@ -24,14 +24,11 @@ const ToolsPanel = (
 		panelContext,
 		resetAllItems,
 		toggleItem,
-		className,
+		...toolsPanelProps
 	} = useToolsPanel( props );
 
-	// Props are not directly passed through to avoid exposing Grid props
-	// until agreement has been reached on how ToolsPanel layout should be
-	// handled.
 	return (
-		<Grid columns={ 2 } className={ className } ref={ forwardedRef }>
+		<Grid { ...toolsPanelProps } columns={ 2 } ref={ forwardedRef }>
 			<ToolsPanelContext.Provider value={ panelContext }>
 				<ToolsPanelHeader
 					label={ label }
```
</details>

aaronrobertshaw replied:
A PR to apply the above changes can be found in: https://github.com/WordPress/gutenberg/pull/36428

ciampo reviewed (commented):
A bit late to the party, but — first of all, kudos to everyone for shipping this PR! It's great to see us moving away from the `single-column` className.

Regarding the next steps, I think that we may be able to make a better decision after getting a bit more experience in using ToolsPanel in more place sin Gutenberg — what do folks think?

I also left a couple of minor comments, possibly requiring a small follow-up PR.
